Public/Src/Engine/Processes/FileAccessManifest.cs

```diff
@@ -504,17 +728,86 @@ public ArraySegment<byte> GetPayloadBytes(FileAccessSetup setup, MemoryStream st
                 WriteInjectionTimeoutBlock(writer, timeoutMins);
                 WriteTranslationPathStrings(writer, DirectoryTranslator);
                 WriteErrorDumpLocation(writer, InternalDetoursErrorNotificationFile);
-                WriteFlagsBlock(writer);
+                WriteFlagsBlock(writer, m_fileAccessManifestFlag);
                 WriteExtraFlagsBlock(writer, FileAccessManifestExtraFlag.None);
                 WritePipId(writer, PipId);
                 WriteReportBlock(writer, setup);
                 WriteDllBlock(writer, setup);
-                WriteManifestTreeBlock(writer);
+                if (!IsManifestTreeBlockSealed)
+                {
+                    WriteManifestTreeBlock(writer);
+                }
+                else
+                {
+                    Contract.Assert(m_sealedManifestTreeBlock != null);
+                    writer.Write(m_sealedManifestTreeBlock);
+                }
 
                 return new ArraySegment<byte>(stream.GetBuffer(), 0, (int)stream.Position);
             }
         }
 
+        /// <summary>
+        /// Serializes this manifest.
+        /// </summary>
+        public void Serialize(Stream stream)
+        {
+            Contract.Requires(stream != null);
+
+            using (var writer = new BinaryWriter(stream, Encoding.Unicode, true))
+            {
+                WriteTranslationPathStrings(writer, DirectoryTranslator);
+                WriteErrorDumpLocation(writer, InternalDetoursErrorNotificationFile);
+                WriteFlagsBlock(writer, m_fileAccessManifestFlag);
+                WritePipId(writer, PipId);
+                WriteChars(writer, m_messageCountSemaphoreName);
+
+                // The manifest tree block has to be serialized the last.
+                if (!IsManifestTreeBlockSealed)
+                {
+                    WriteManifestTreeBlock(writer);
+                }
+                else
+                {
+                    Contract.Assert(m_sealedManifestTreeBlock != null);
+                    writer.Write(m_sealedManifestTreeBlock);
+                }

```

Comment:
You could move this whole `if (!IsManifestTreeBlockSealed) { ... } else { ... }` block to `WriteManifestTreeBlock` so that you don't have to repeat it both here and in `GetPayloadBytes()`

Reply:
it looks like you've already done that, so maybe this `if` (and the one in `GetPayloadBytes`) is actually unnecessary

Reply:
agree.

---
In reply to: [276464305](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276464305) [](ancestors = 276464305)

Reply:
Unfortunately not. GetManifestTreeBytes does serialize the manifest if manifest is not sealed. For testing, this method is appropriate, but it doesn't seem we can use that for serialization and deserialization.

---
In reply to: [276464697](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276464697) [](ancestors = 276464697)